Append logmein param to links on click

lsl opened the pull request:
#### Changes proposed in this Pull Request

* Extension of work in #52909 
* Adds onClick intercept for clicks on links, if the link is pointing at a simple site with a mapped domain it swaps out the url for one including a `?logmein=direct` param.
* Adds replaces urls in the same way for `navigate` calls

#### Testing instructions
![Screenshot(18)](https://user-images.githubusercontent.com/811776/123238446-d6d74000-d521-11eb-8ccd-751caefa9e9b.png)

* Find links in calypso that link out to simple sites with mapped domains.
* "Visit site" on the "My Home" page works for this, so do the domain pages under Plans > Domains
* If the domain is a mapped domain against a simple site,
  * You should see a couple of effects:
    * It should take you through the redirect login flow and land you logged in to the site. (check wordpress_logged_in cookie is set against the domain directly or [check the redirect flow](https://chrome.google.com/webstore/detail/link-redirect-trace/nnpljppamoaalgkieeciijbcccohlpoh?hl=en))
    * After clicking one of these links the links `href` will have been updated (if your link has target=_blank).
* Any other site types (atomic, jetpack) should work as normal.

Unit tests:
```
yarn test-client client/lib/logmein
```

Related to #52741 / #52740 / #52763

lsl committed:
Add logmeinUrl as lib

lsl committed:
Add getSitesMappedDomains selector,

and use it with logmeinUrl to apply logmein params to domain links.

lsl committed:
Add tests back

lsl committed:
Test fixes, mv

lsl committed:
Rename mapped domains selector to be more specific, shore up logmein https, direct flag value

lsl committed:
Default logged in and direct, test fixes

lsl committed:
delete me - config/development.json

lsl committed:
Externalize allowed url check, move to lib

lsl committed:
wip: attachLogmein

lsl committed:
Intercept for clicks and navigate

lsl commented on client/my-sites/domains/domain-management/list/index.jsx:
```diff
@@ -53,6 +53,7 @@ import InfoPopover from 'calypso/components/info-popover';
 import ExternalLink from 'calypso/components/external-link';
 import HeaderCart from 'calypso/my-sites/checkout/cart/header-cart';
 import AddDomainButton from 'calypso/my-sites/domains/domain-management/list/add-domain-button';
+import { logmeinUrl } from 'calypso/lib/logmein';
```
modifying this file is no longer required - links are handled automagically

lsl commented on client/lib/logmein/test/index.js:
```diff
@@ -0,0 +1,62 @@
+/**
```
needs new tests

lsl commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,129 @@
+		)
+		.filter( Boolean );
+}
+
```
Dont know if I like these smaller functions, but it cleans up the distinction between being logged in (most common) or logged out (rarely needed if ever?)

lsl replied:
Logmein feature may need ammending to handle packing of url hash values. Remote login already supports this but it was omitted from logmein.

Permalinks on [comment approval screen](http://calypso.localhost:3000/comment/c3polikes.blog/73) link to comments like this: 

https://c3polikes.blog/2021/04/06/sedate-tuna-beg-daily/comment-page-1/#comment-73

This actually seems to work (hash follows redirects but this might be a Firefox thing or already logged in (wpcom or site) thing)

lsl replied:
Ignoring hash issue until its a confirmed problem. Removed these extra functions, only supporting logged in usage for now.

lsl commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,129 @@
+	return url;
+}
+
+let allowedSites: string[] = [];
```
Is there a better place/way to store this response for use by the navigate wrapper and onclicker?

jsnajdr replied:
You can replace the `wpcom.me().sites()` request by reading from Redux store and relying on someone else issuing the request for you. We currently always issue the request in the main layout except some Jetpack Connect routes.

Your click handler can do something like:
```js
function onClick( event ) {
  const link = event.target.href;
  const sites = Object.values( getSitesItems( reduxStore.getState() ) );
  if ( sites.some( site => link.startsWith( site.URL ) ) ) {
    event.target.href = addLogmeinQuery( link );
  }
}
```
That will use the latest known sites list on each click.

lsl replied:
Thanks! Exactly what I was looking for! Do you know where this main layout request happens? I could not track it down (though could see the request)

jsnajdr replied:
In `client/layout/index.jsx` we render a `<QuerySites allSites />` component which does the request on mount. And there are several places that dispatch the `requestSites` Redux action to refresh the sites list, e.g., after accepting an invite to a new one etc.

p-jackson replied:
Will we need to attach `logmein` query params in situations where the Calypso redux store doesn't exist? For example the ETK plugin or the notifications popup iframe?

If so, then what about if `attachLogmein()` accepted an `allowedSites` parameter? If it's missing then `attachLogmein()` will have to make the request itself. If it's provided then it doesn't have to make the extra request.

When `attachLogmein()` is called in `client/boot/common`, since it has access to the redux store right there, it can pass in the allowed sites.

p-jackson replied:
But since this is in `client/lib/logmein`, perhaps it's only meant to be used inside of Calypso.

lsl replied:
> Will we need to attach logmein query params in situations where the Calypso redux store doesn't exist?

[Yes, see editor iframe link comments re ETK.](https://github.com/Automattic/wp-calypso/pull/53443#discussion_r647925730)

Other situations less likely but I would like the option to leverage at least some of the same code.

> But since this is in client/lib/logmein, perhaps it's only meant to be used inside of Calypso.

I started out in `./package` land but it over complicated the MVP so I moved to `lib` knowing full well I'd probably need to move it to a package once I had enough working to ping you for better ideas :smile:

lsl replied:
Tackling this in a follow up.

lsl commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,129 @@
+	}
+	return newurl.toString();
+}
+
```
Can we find if a site is a mapped simple site? I don't trust this exclusion approach. May need to extend this to atomic soon though so? :shrug:

lsl replied:
@niranjan [mentions](https://github.com/Automattic/wp-calypso/pull/52909#discussion_r646360802) using:

```
! isJetpackSite( state, siteId ) || ! isAtomicSite( state, siteId )
```

It looks like these are covered below (`isJetpackSite` [checks](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/sites/selectors/is-jetpack-site.js#L20) `site.jetpack`, `isAtomicSite` [checks](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/selectors/is-site-automated-transfer.js) `site.options.is_automated_transfer`.

It'd be good if anyone can chime in about these other options I'm checking. I just picked things out of the site object/options that might point at products we don't cover with logmein support.

- assuming `vip` is `vip`
- assuming `jetpack` is `jetpack`
- what's the difference between `is_wpcom_atomic` and `is_automated_transfer`? 
- assuming `domain_only` is a dns only setup, ie. server not going to be wpcom
- assuming `is_redirect` is somewhat the same as `domain_only`
- assuming `is_wpcom_store` is woo and therefore atomic
- assuming `is_wpforteams_site` is p2, @josephscott do you know if p2 supports mapped domains and therefore should be covered by logmein (i.e. remove this restriction and let it get picked up by is_mapped_domain)

I tried to find context on specifics but mgs turns up a few blanks.

jsnajdr replied:
Simple site is one that is not Jetpack. Atomic sites are a subset of Jetpack sites, and have the `options.is_automated_transfer` flag.

The `is_wpcom_atomic` flag is for sites that run on a newer version of Atomic infrastructure, I think you don't need to check that here.

lsl replied:
Are VIP sites Jetpack?

Ah so is_wpcom_atomic is subset of is_automated_transfer. Probably don't need this for now but may come in handy later if/when we support logmein on atomic.

jsnajdr replied:
> Are VIP sites Jetpack?

I'm not sure as I have zero experience with VIP, but I think a VIP site can be technically anything. Simple site, Atomic site, other hosting (vipgo) with or without Jetpack...

niranjan-uma-shankar replied:
> ! isJetpackSite( state, siteId ) || ! isAtomicSite( state, siteId )

I'm looking at this again and realize made a mistake. The condition after the OR is not required, just  `! isJetpackSite( state, siteId )` is what I was going for to cover simple sites. 

Going to the checks below, looks like `! site.jetpack` will exclude atomic too, so I guess we don't need separate checks for atomic. `! site.options.is_domain_only` and `site.options.is_mapped_domain` make sense in addition to the jetpack check. I think these three should cover the case for VIP, WP for teams etc. since if it's a mapped domain site that is not Jetpack, then `logmein` wouldn't technically fail. 

> do you know if p2 supports mapped domains 

@lamosty Does P2 product support custom domains yet?

lsl replied:
> I'm looking at this again and realize made a mistake. 

I didn't notice but I understood what you meant anyway :laughing: 

> then logmein wouldn't technically fail.

The failure mode is what scares me the most about this, with the `direct` mode of logmein if you're not logged in to wordpress.com during the redirect flow you'll land on a blank white screen.

This probably needs fixing with some kind of fallback as I suspect safari ITP might break this flow at some point by treating cookies in redirect hops as third party (blocked).

naxoc replied:
> Does P2 product support custom domains yet?

No, we don't support them just yet. At least not officially. There are a couple of them out there from early days before we ran into problems with Safari and them, but we don't support them.

lsl replied:
Thanks @naxoc, I pinged @lamosty and got the following: 

- Not currently supported
- Plan to support them once 3p cookie problems are sorted
- They're all simple sites

Since they're all simple sites (and I assume going to stay that way) I'm going to treat them as valid allowing them to be logmein'd if they're also mapped. This should be a noop until we start supporting mapped domains for p2. Once we do support them, logmein redirect links should just work like any other simple site.

lsl committed:
Undo domains page change, using link injection now

lsl committed:
Refactor, replace getlink with closest

lsl commented on client/controller/index.web.js:
```diff
@@ -47,6 +48,8 @@ export const ProviderWrappedLayout = ( {
 	const userLoggedIn = isUserLoggedIn( state );
 
+	attachLogmein( userLoggedIn );
+
```
This is the source of the duplicated /me/sites hits, there is probably a better place to do this, boot/common maybe?

jsnajdr replied:
Yes, you want to move this to `client/boot/common`, near the `detectHistoryNavigation.start()` call. Which also adds a global DOM listener. Another global DOM listener is installed in `accessibleFocus()` in the same module. But this one runs before the Redux store is created.

At its current location, in `ProviderWrappedLayout`, the listener would be added again and again on each React rerender.

lsl replied:
Thanks, will do tomorrow!

lsl replied:
Dropped this into the middlewares function, not middleware but it seems like the best place for this (there is similar non middleware stuff happening later in the method).

lsl commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,128 @@
+// Used as placeholder / default domain to detect when we're looking at a relative url
+const INVALID_URL = `https://__domain__.invalid`;
+
+export function logmeinUrl( url: string, sites: any, isWPComLoggedIn = true ): string {
```
Can potentially drop this function now - no longer used, keeping it round while we test, there may be situations where we can't use the click handler

lsl replied:
Reworked

lsl commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,128 @@
+			} );
+	}
+}
+
```
This or the allow list may need to get more complicated:

The "View" post link (e.g. in the dropdown menu (...) of a post listed on http://calypso.localhost:3000/posts/c3polikes.blog ) loads the previewer using the `*.wordpress.com` domain, this is fine for the previewer but the previewer then links to this domain relying on redirects.

This code treats this situation as a no match and doesn't append the logmein param.

Fix probably involves doing a mapping step before the indexOf check.

jsnajdr replied:
In that case you need to check `site.options.unmapped_url`, too. The `*.wordpress.com` variant is there. While `site.URL` contains the primary mapped domain.

niranjan-uma-shankar replied:
> The "View" post link (e.g. in the dropdown menu (...) of a post listed on http://calypso.localhost:3000/posts/c3polikes.blog ) loads the previewer using the *.wordpress.com domain, this is fine for the previewer but the previewer then links to this domain relying on redirects.
This code treats this situation as a no match and doesn't append the logmein param.

Adding to this, I also found that the logmein doesn't work if I click on links in the editor:

<img width="1661" alt="Screenshot 2021-06-08 at 7 23 04 PM" src="https://user-images.githubusercontent.com/1269602/121213289-27e30500-c88f-11eb-9ec7-58298e3e19e6.png">

niranjan-uma-shankar replied:
The intercept also didn't work on the themes page, in the pop-up that appears after activating a theme:

<img width="467" alt="Screenshot 2021-06-08 at 7 31 20 PM" src="https://user-images.githubusercontent.com/1269602/121215724-4f3ad180-c891-11eb-9c90-ac72cf696a81.png">

lsl replied:
Editor won't work because its iframed to the *.wordpress.com site (Even on mapped domains because similar 3rd party cookie reasons) 

Fix needs to be implemented in [ETK](https://github.com/Automattic/wp-calypso/tree/trunk/apps/editing-toolkit) or move this to a package ETK can use. (why I started there first go round but it overcomplicated getting it working)

It can be done in a serverside plugin for simple sites only but I want this to work for atomic eventually so we need ETK for this.

lsl replied:
Thinking more: might be worth deploying this behind feature flag first and iterating on a solution with ETK, already a lot happening here. Potentially need to move a lot of files around though :thinking: let me see what an ETK fix might look like ~today~ tomorrow.

jsnajdr replied:
> The intercept also didn't work on the themes page, in the pop-up that appears after activating a theme:

That's because the "Visit site" is not a link, but a button that does `window.open( url )` in an `onClick` handler. We can't possibly intercept that, except maybe patching the `window.open` function. The best solution would be to convert that button to `<a href>`. That improves accessibility, too.

jsnajdr replied:
> Editor won't work because its iframed to the *.wordpress.com site

One solution for that: intercept `<a href>` clicks in the editor iframe, and then send them to the parent window as an `openLink` message. The parent window has access to the sites list and can do the navigation properly.

I vaguely remember we do this technique already somewhere else -- maybe in post preview iframes? Previews do a lot of `postMessage` communication with their parent frame.

lsl replied:
> That's because the "Visit site" is not a link, but a button that does window.open( url ) in an onClick handler.

Didn't realize this was a button. Agree with changing to `<a>` but that will still require some kind of fix, not working example: the links in the sidebar permalink section fail to get the logmein treatment. I'll take a closer look at post preview iframes to see what we can do with postMessage for that.

lsl replied:
Added premapping of unmapped_url's onto their mapped domains.

Todo: postmessage for editor or etk fix.

lsl replied:
Update: "View Site" link on theme page now caught and replaced.

lsl replied:
Update: "Visit Site" in page/post preview is updated on right click and left click.

lsl replied:
Tackling this in a follow up.

jsnajdr commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,128 @@
+	} else {
+		window.location.href = url;
+	}
+}
```
I would separate the two responsibilites that this function has:
- optionally appending the `logmein` query arg
- doing the actual navigation

Keeping only the first one, and letting the caller handle the second one.

lsl replied:
I assume you mean like call logmeinUrl (or whatever) from the navigate lib, good idea, much better way to go.

lsl replied:
Cleaned up logmeinUrl a bit along side the redux rework, managed to drop this method completely using logmeinUrl.

jsnajdr commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,128 @@
+		if ( allowedSites.indexOf( url.hostname ) !== -1 ) {
+			url = appendLogmeinDirect( url );
+			console.log( 'intercepted', link.href, 'replaced', url.toString() );
+			link.href = url.toString();
```
This changes the actual `href` attribute of the `a` element and the default browser handler will use that to navigate. That's cool and smart 🙂 In some edge cases it might do that repeatedly though, on each click.

lsl replied:
It probably will, though I think I used the right `set` call on URL to make that safe (will double check)

I was thinking I could clear the value after but its probably overkill.

p-jackson replied:
This is a novel approach, I hadn't thought there'd be a way of fixing all the links in one hit like this does.

But will this approach break the right-click to open in a new tab/window experience? Perhaps there's a way of getting a click handler to fire even when the link is right-clicked so we can adjust the `href` in that case too. But then there's probably a keyboard driven way to open in a new tab/window too. These are downsides that the `localizeUrl()` approach doesn't have (even though it'd mean peppering a call to `logmeinUrl()` all through the code.

lsl replied:
Right click open doesn't work with this PR as is, [should](https://stackoverflow.com/a/4236294) be able to detect right click events using [`contextmenu`](https://developer.mozilla.org/en-US/docs/Web/API/Element/contextmenu_event) events whether these happen at the right time needs checking.

Keyboard `tab` to focus + `enter` does.

Looking up options it [might be possible](https://stackoverflow.com/questions/3522090/event-when-window-location-href-changes) to track the href value before navigation using [`MutationObserver`](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver).

It might also be worth checking if there is any undue performance hit using MutationObserver to watch and check links as they're added to the dom. I've done similar things in the past but it was expensive. I wasn't using mutation observers though so maybe it would work better?

p-jackson replied:
TIL about the `contextmenu` event 👍

lsl replied:
Added contextmenu event handler, seems to fire on right click so you don't need to perform any action in the context menu and the link still gets updated.

lsl replied:
> In some edge cases it might do that repeatedly though, on each click.

Also added a `seen[host]` check so we'll only set it once per click, right click always applies as you need to ensure the user visits the link before you know its safe to stop adding it.

niranjan-uma-shankar commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,128 @@
+			.me()
+			.sites()
+			.then( ( sites: any ) => {
+				allowedSites = logmeinAllowedUrls( sites.sites );
```
Should we relax the `allowedSites` list to any site that points to WP.com? Users could also navigate to other WPCOM sites from Reader, and I think it would be helpful to have those links intercepted and appended with `logmein` too. Probably using something like `pointsToWpcom` as is used [here](https://github.com/Automattic/wp-calypso/blob/6b9cee4164c4fbba7a9a667b031601509db1c1a4/client/my-sites/domains/components/domain-warnings/index.jsx#L167). What do you think?

lsl replied:
I was wondering about that variable, looking it up, its true for any a8c ip, VIP, atomic, wpcom, and optionally some proxies so probably a no go for simple site identification.

Potentially an option in the future, I wonder if VIP is going to need the same magic we're cooking up for atomic? (asked here p1623207859067300-slack-C01M493LTGD)

Re navigation from the reader I didn't think about this use case but its def something we should look at. We'd need to do a site lookup for every domain displayed, that might be a bit heavy and might have timing issues on 3g.

The reader is potentially the only calypso context where this happens, its worth looking to see if we can update the endpoint response to return site info along with the post links (or logmein urls preloaded) and then do an extra change for the reader links specifically.

niranjan-uma-shankar replied:
Agree with everything you said.

> The reader is potentially the only calypso context where this happens, its worth looking to see if we can update the endpoint response to return site info along with the post links (or logmein urls preloaded) and then do an extra change for the reader links specifically.

Yes, this seems a reasonable approach to try in a future iteration.

niranjan-uma-shankar commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,128 @@
+
+	if ( link ) {
+		let url = new URL( link.href, INVALID_URL );
+		if ( allowedSites.indexOf( url.hostname ) !== -1 ) {
```
Not really something for this PR, but I was wondering if for a future iteration, we could add a check here to ensure that we redirect to `/?logmein=direct` only if the user has not already gone via a logmein link elsewhere in Calypso. If the user has previously navigated to a logmein link, then no point taking them through it again, and we can save an extra redirect for https://domain-1.tld/?logmein=direct -> https://domain-1.tld. 
Probably for this, we need to track the logmein clicks in session storage? Need to ascertain if it's worth the effort though - if all it does is shave off a few nanoseconds in page load for an infrequent navigation, then it's probably not worth the effort.

lsl replied:
This is potentially doable, would just need to be concerned about expiries / cookie clears. Short timeout on it might be enough.

Would tracking this in memory(variable state) be good enough? There would still be a lot of redirects but we'd stamp out the most excessive scenarios and be super safe around cookie clear situations.

lsl replied:
Added in memory check to track these.

niranjan-uma-shankar commented:
I hope you don't mind that I did a rebase since I wanted the newly launched Calypso oauth feature.

niranjan-uma-shankar commented:
Updating that I also tried this in Safari with "Prevent cross site tracking" enabled and the intercepts worked fine.

p-jackson commented on client/lib/navigate/index.ts:
```diff
@@ -3,6 +3,11 @@
+/**
+ * Internal dependencies
+ */
+import logmeinNavigate from 'calypso/lib/logmein';
```
```suggestion
import { logmeinNavigate } from 'calypso/lib/logmein';
```

p-jackson commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,128 @@
+			.sites()
+			.then( ( sites: any ) => {
+				allowedSites = logmeinAllowedUrls( sites.sites );
+				document.addEventListener( 'click', logmeinOnClick, false );
```
I wonder if this should use `true` for the `useCapture` parameter. That way we'd get a chance to modify the `href` before any other event handlers run, and perhaps we'd want them to see the updated url.

Just a thought, not sure if that's the right thing to do.

lsl replied:
That sounds potentially useful TIL, will check it works, there might be an issue with the way the closest lookup works in with this.

lsl replied:
Capturing doesn't seem to work, I suspect the bubbling is happening in the opposite order we expected. Leaving this as false for now.

p-jackson commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,128 @@
+}
+
+function logmeinOnClick( event: MouseEvent ) {
+	const link = ( event.target as HTMLElement ).closest( 'a' );
```
I'm wondering what the performance implications of adding a tree traversal to every single mouse click in the app will be. Perhaps the nature of mouse clicks mean this'll be fine (it's not like dragging or typing, the user can probably tolerate a few milliseconds delay for a click). Just thought it was worth bringing up since this is very global.

lsl replied:
Definitely worth bringing up.

The traversal should be linear time so not too concerning. I'm guessing `closest` [looks like this](https://github.com/Automattic/wp-calypso/pull/53443/commits/1585505622ab70fdd7db602aa7ea16e8a2500dfb#diff-d025c6a21229fe69836222916356ee8a2c395861252a82ff88e97b9facc70a4bL95) chaining up `.parent`. I would like to back this up if we can though the globalness of this is what concerns me the most.

Lets discuss this [above here](https://github.com/Automattic/wp-calypso/pull/53443#discussion_r647994782) as watching a mutation observer for links and appending preemptively might be a better option.

jsnajdr replied:
> I'm wondering what the performance implications of adding a tree traversal to every single mouse click in the app will be.

The performance impact should be negligible, because this tree traversal already happens many many times on every event, and adding one more is unlikely to be a concern.

React itself adds a global `click` listener to the `document` or the React root element. When a click happens on e.g., a `<button>`, React is traversing the DOM tree from the `<button>` element up to the React root and dispatches synthetic events on the way.

The `page.js` router installs a global `click` handler where it intercepts all clicks on `<a href>` links and handles them using `page.js` code instead of letting the browser do a native navigation with full reload. That means that a click handler almost identical to the logmein one is already active in Calypso.

The browser itself does a DOM traversal from the root `window` element to the target one (capturing) and back to root again (bubbling), finding and invoking event handlers on the way.

Browser extensions can install their own intercepting event handlers. For example, my Firefox Devtools tell me that the 1Password extension is watching all `click`s on my `<button>` elements, using a `closest( 'button' )`-like traversal.

lsl replied:
Too much TIL in this PR. :bow: 

Good to know this is a rather well accepted/safe approach to attach click events in a generalized way.

p-jackson replied:
Great points 👍

lsl committed:
Fix import of logmeinNavigate

Co-authored-by: Philip Jackson <p-jackson@live.com>

lsl committed:
Checking href exists, probably just being over cautious

lsl committed:
Split valid site check method, remove consoles

lsl committed:
Move attach to boot/common + switch to redux store for site opts

lsl committed:
Navigate rework and other shuffling

lsl committed:
fix isEnabled - autofix imports, also move user logged in check to
lib... not sure how I like this

lsl committed:
Map unmapped urls to mapped domains before checking for logmein applicability

lsl committed:
more type

lsl committed:
namefix

lsl committed:
Only apply logmein to each host click once

lsl committed:
Handle on right click

jsnajdr commented on client/boot/common.js:
```diff
@@ -410,6 +411,11 @@ const setupMiddlewares = ( currentUser, reduxStore ) => {
 		} );
 	}
+
+	if ( config.isEnabled( 'logmein' ) && currentUser && currentUser.get() ) {
```
The logged-in check should be reading from Redux: `isUserLoggedIn( reduxStore.getState() )`. @tyxla is working on removing the `lib/user` helper object.

lsl replied:
Thanks for the heads up

jsnajdr commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,121 @@
+	}
+
+	// logmein=direct only works for logged into wordpress.com users
+	if ( ! userUtils.isLoggedIn() ) {
```
@tyxla just removed this helper in #53621. Let' use Redux here, too.

jsnajdr commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,121 @@
+	document.addEventListener( 'contextmenu', logmeinOnRightClick, false );
+}
+
+const seen: Record< Host, boolean > = {};
```
Another way to do this would be to add a `data-logmein` attribute to the link and check it before changing.

1Password, for example, stores its state data also directly to DOM, in attributes like `data-com.agilebits.onepassword.user-edited`.

lsl replied:
That would be nicer but will only catch repeat clicks on the same link, clicks on other links with the same host won't be caught.

Not sure if it matters either way (I don't think it matters much). I'm going to leave this as is for now but do voice any concerns if you think its worth changing.

jsnajdr commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,121 @@
+	const sites = Object.values( getSitesItems( reduxStore.getState() ) );
+
+	// Replace unmapped url usage with the mapped hostname
+	const hostmap = unmappedToMapped( sites );
```
This creates the `hostmap` on every `logmeinUrl` call, which is inefficient. I think the simplest `sites.find( site => ... )` will be hard to beat in practice.

jsnajdr replied:
And why do you replace `*.wordpress.com` links with the mapped domain? Keeping them `*.wordpress.com` should be OK.

lsl replied:
Updating to use the .find method instead.

---

*.wordpress.com urls are used in the preview window and potentially other places. The code in these places is relying on the fact the mapped domains will do a server side redirect. This redirect won't have ?logmein attached so the user will end up on the page potentially logged out when we really want them to be logged in.

It's better if we fix these urls to use the mapped domain directly but I'm not sure how to track all usages down.

lsl replied:
> And why do you replace *.wordpress.com links with the mapped domain? Keeping them *.wordpress.com should be OK.

Unresolving this as I think I need to justify this a bit better there could be good reasons for linking to the unmapped domain instead of the mapped domain - the domains list for example should rely on server side redirects instead of client side rewrites.

Justification needed is to figure out how many places we do this and if we can't just update them to use the primary domain instead of the unmapped url.

lsl replied:
Removed this.

lsl committed:
Use redux state for login check in boot/common

lsl committed:
Replace userUtils usage with redux store state

lsl committed:
Enable logmein on p2 sites

lsl committed:
Replace hostmap with find props: @jsnajdr

lsl committed:
Refactor + tests

- Refactor store to be passed in at call time to help with testing
- Update tests for logmein, add new ones after previous changes and new
state requirements

lsl committed:
Revert back to encap store, update tests

Forgot about navigation usage not having access to the redux store

lsl committed:
Ensure redux store set

lsl committed:
Cleaning up doc, add test for edge case protocol-less edgecase

lsl commented:
@scinos / quality-squad ignore the ping, I pushed a playwright test now moved to a draft PR #53979

lsl committed:
Remove unmapped url rewrites - this seems unsafe

aneeshd16 commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,116 @@
+	if ( link && link.href ) {
+		// Only apply logmein to each host once
+		const host = new URL( String( link.href ), INVALID_URL ).host;
+		if ( ! seen[ host ] ) {
```
This breaks if the link to the same host is re-rendered. Steps to reproduce:
1. Go to the domains management page (/domains/manage)
2. If the custom domain is not primary, set it as primary, else ignore this step.
3. Under the "Primary domain" section, click on the link to the domain.
4. A new tab opens and you are logged in to that domain. ✅ 
5. In the same tab, clear cookies for that domain.
6. Go to the tab with the domain management page.
7. Set the *.wordpress.com subdomain as primary.
8. Set the custom domain as primary again.
9. Now, click on the link to the domain again.
10. You'll not be logged in when the site for that domain opens.

I guess this will also break when there are multiple links pointing to the same host.

After a bit more testing, there is an easier way to reproduce this:
1. Repeat steps 1-6 above
2. In the second section, click on the domain to open the settings for that domain.
3. Click on the link to the domain.
4. You'll not be logged in when the site for that domain opens.

The `seen` object is preserved between page redirects too, so there are other ways to reproduce this too. I missed the previous discussions here, so I'm curious to know its purpose (apart from being a performance optimization)?

lsl replied:
> This breaks if the link to the same host is re-rendered. Steps to reproduce:

This is the intention of the seen variable, the idea is to prevent people going through the logmein redirect flow more often than they need to. Once they're logged into the host they won't need to be logged in again for a while. Clearing cookies in between tabs is unlikely enough for it not to be an issue.

You can probably make the argument this is creating a flaky user experience and I'd agree but I think the trade off is worth it to prevent some additional redirects.

> The seen object is preserved between page redirects too, so there are other ways to reproduce this too. I missed the previous discussions here, so I'm curious to know its purpose 

Can you elaborate on this? My understanding is it being an in memory variable, shouldn't be cached between reloads - but pushstate navigation probably will survive.

> so I'm curious to know its purpose 

There was an earlier conversation in this PR with Niranjan about it but I can't track it down (rebases). There is no way to know if a user is logged in to a domain so we take this approach to reduce redirects, there will still be redirects that aren't needed but this is easier than tracking logged in state everywhere.

niranjan-uma-shankar replied:
> There was an earlier conversation in this PR with Niranjan about it but I can't track it down 
It was on [this thread](https://github.com/Automattic/wp-calypso/pull/53443#discussion_r647586195).

aneeshd16 replied:
> This is the intention of the seen variable, the idea is to prevent people going through the logmein redirect flow more often than they need to. Once they're logged into the host they won't need to be logged in again for a while. Clearing cookies in between tabs is unlikely enough for it not to be an issue.

Makes sense! It slipped my mind that the link is modified on click and not on render, so we are good here!

> There is no way to know if a user is logged in to a domain so we take this approach to reduce redirects, there will still be redirects that aren't needed but this is easier than tracking logged in state everywhere.

Thanks! This does clear it up.

aneeshd16 commented on client/lib/logmein/index.ts:
```diff
@@ -49,14 +49,7 @@ export function logmeinUrl( url: string ): string {
 
 	// Replace unmapped urls with their mapped url counterparts (*.wordpress.com => mapped.example.com)
```
This comment might need to be removed too :)

niranjan-uma-shankar reviewed (approved):
I tested the following scenarios:

✅ Calypso links for public, private, and coming soon mapped domain simple sites, logmein redirect worked
✅ Logmein redirect does not happen for *.wordpress.com subdomain sites
✅ Logmein redirect works for .blog subdomain sites like example.food.blog
✅ Logmein redirect does not happen for atomic sites
✅ Logmein redirect does not happen for Jetpack sites
❌ Links clicked in Reader do not get logmein redirects(mentioned earlier here and we discussed tackling it separately, just noting here for completion sake)
❌ On private simple sites, clicking the site preview from sidebar opens the website instead of opening the preview iframed in Calypso. [Here's a gif](https://cloudup.com/cdRkIBQs_Dn). 
**UPDATE**: This happens on trunk branch too, so I think it's just something to do with calypso .localhost domain with 3PC disabled and not related to this PR.
❌ I couldn't get to make this PR work with the calypso.live link, but it worked with calypso.localhost. I just wanted to call this out, maybe it's not a big deal.

Q: Do you think we should enable this for staging environment too? That could get more eyeballs from a12s who use WP.com proxied. 

Overall, LGTM so I'm approving it from my end. I see @aneeshd16 has some comments, so those might need to be checked.

lsl committed:
Auxclick support

lsl commented:
> x Links clicked in Reader do not get logmein redirects(mentioned earlier here and we discussed tackling it separately, just noting here for completion sake)

:+1: 

> x On private simple sites, clicking the site preview from sidebar opens the website instead of opening the preview iframed in Calypso. [Here's a gif](https://cloudup.com/cdRkIBQs_Dn).
> **UPDATE**: This happens on trunk branch too, so I think it's just something to do with calypso .localhost domain with 3PC disabled and not related to this PR.

:+1: unrelated to this change, see https://github.com/Automattic/wp-calypso/blob/trunk/client/components/web-preview/content.jsx#L243 with 3pc disabled the preview fails to load triggering a less than ideal fallback.

> x I couldn't get to make this PR work with the calypso.live link, but it worked with calypso.localhost. I just wanted to call this out, maybe it's not a big deal.

Try adding `?flags=logmein` to the url - I'll update wpcalypso.json

> Q: Do you think we should enable this for staging environment too? That could get more eyeballs from a12s who use WP.com proxied.

I'd like to hold off on a call for testing until we have the editor sorted out.

> Overall, LGTM so I'm approving it from my end. I see @aneeshd16 has some comments, so those might need to be checked.

:+1:

lsl committed:
Enable logmein on wpcalypso / calypso.live

niranjan-uma-shankar commented:
> Try adding ?flags=logmein to the url - I'll update wpcalypso.json

Oh yea, makes sense it didn't work earlier since wpcalypso environment was not enabled.

> I'd like to hold off on a call for testing until we have the editor sorted out.

Sounds good. Do you mind creating issues for these for tracking purpose? I think there are these three to be figured out:
1. Editor links
2. View site link in the [preview frame](https://github.com/Automattic/wp-calypso/pull/53443#discussion_r655053643)
3. Links to other WP.com sites in the Reader

Also, there's a unit test failing.

jsnajdr commented on client/lib/logmein/index.ts:
```diff
@@ -0,0 +1,138 @@
+		.filter( Boolean );
+	if ( allowedHosts.indexOf( newurl.host ) === -1 ) {
+		return url;
+	}
```
I would write this with `.find` or `.some`:
```js
const isAllowed = sites.some( ( site ) => {
  return (
    isValidLogmeinSite( site ) &&
    new URL( site.URL, INVALID_URL ).host === newurl.host
  );
} );
```

jsnajdr reviewed (approved):
Looks good and works well :+1:

lsl committed:
Test fix for auxclick change

lsl committed:
Array loop optimization props: jsnajdr

scinos commented:
I think this PR breaks on IE11.

I was testing different branch and I saw 

![image](https://user-images.githubusercontent.com/975703/127109140-7f25a9f6-2486-4f9a-b44d-d22c23c972fe.png)

Opening the bundled file and prettyfing it shows the errors comes from

```
			function u(e) {
				const t = e.target.closest("a");
				if (t && t.href) {
					const e = new URL(String(t.href), i).host;
					d[e] || (t.href = c(t.href), d[e] = !0)
				}
			}
```

In the file `./lib/logmein/index.ts`. Likely this line https://github.com/Automattic/wp-calypso/pull/53443/files#diff-d025c6a21229fe69836222916356ee8a2c395861252a82ff88e97b9facc70a4bR105

jsnajdr commented:
Good catch @scinos! I thought that `.closest` is polyfilled by Core JS 3, but it isn't. We're using `.closest` quite frequently in both Calypso and Gutenberg, so maybe both projects would benefit from adding a polyfill? I googled this one: https://vanillajstoolkit.com/polyfills/closest/

I'm not sure if it's worth it as deprecation of IE11 is happening. Cc @sgomes

sgomes commented:
Oh wow, I had no idea that `.closest` wasn't being polyfilled by `core-js` nor `babel`. The level of inconsistency in these things can be very annoying.

Thank you for mentioning me, @jsnajdr, and good work on uncovering this, @scinos!

jsnajdr commented:
I just found that the Core `wp-polyfill` script includes a `wp-polyfill-element-closest` submodule. So we have a Calypso-only problem here. I'm going to add the polyfill.
